### PR TITLE
Pagination#51

### PR DESCRIPTION
**Pull request by AnnaOstrovska**
## Type of PR
This PR is a feature

## Feature or Bug Description
Pagination of posts by scrolling

## Solution Description
Every time the scroll is near the end of the page, new posts start loading. This happens due to a new request to the server, which indicates a new page number and the number of posts to be downloaded.

## Output Screenshots
Peek 2021-11-18 11-30

### CODE REVIEW

File: src/screens/main/MainScreen.tsx [hunk 1]
```diff
@@ -18,6 +19,40 @@ const MainScreen: React.FC = () => {
...
+
+  const time = 1000;
+  const delay = _.debounce(render, time);
+  window.addEventListener("scroll", delay);
+
+  useEffect(() => {
+    return () => {
+      window.removeEventListener("scroll", delay);
+    };
+  }, []);
```
Comment (irenlian) [inline src/screens/main/MainScreen.tsx:55]: All this scroll-related logic could be moved into the custom hook.

Comment (irenlian) [inline src/screens/main/MainScreen.tsx:55]: `delay` creates each time during rerender.
After each rerender you'll have a new listener added.
After `componentDidMount` only the first listener will be removed.

File: src/screens/main/MainScreen.tsx [hunk 2]
```diff
@@ -18,6 +19,40 @@ const MainScreen: React.FC = () => {
   const posts = useSelector((state: StatePosts) => state.posts.posts);
 
   const [searchTerm, setSearchTerm] = useState("");
+  const [page, setPage] = useState(1);
+  const dispatch = useDispatch();
+  const [fetching, setFetching] = useState(true);
+  let scrollHeight = document.documentElement.scrollHeight;
+  let currentScrollPosition = document.documentElement.scrollTop;
```
Comment (irenlian) [inline src/screens/main/MainScreen.tsx:26]: Each time you call `document` it queries the document. It is consuming, so please minimize these calls.

File: src/screens/main/MainScreen.tsx [hunk 3]
```diff
@@ -18,6 +19,40 @@ const MainScreen: React.FC = () => {
...
+    }
+  }, [fetching]);
+
+  const render = () => {
+    if (scrollHeight - (currentScrollPosition + visibleArea) < 300) {
+      setFetching(true);
+    }
+  };
+
+  const time = 1000;
```
Comment (irenlian) [inline src/screens/main/MainScreen.tsx:47]: This is a constant. SHould not be here

File: src/screens/main/MainScreen.tsx [hunk 4]
```diff
@@ -18,6 +19,40 @@ const MainScreen: React.FC = () => {
...
+  const [page, setPage] = useState(1);
+  const dispatch = useDispatch();
+  const [fetching, setFetching] = useState(true);
+  let scrollHeight = document.documentElement.scrollHeight;
+  let currentScrollPosition = document.documentElement.scrollTop;
+  let visibleArea = window.innerHeight;
+
+  useEffect(() => {
+    if (fetching) {
+      fetch(`https://starforum.herokuapp.com/api/v1/posts?page=${page}&number=5`)
```
Comment (vkravchuk2) [inline src/screens/main/MainScreen.tsx:31]: Нужно заменить на аксиос запрос

File: src/screens/main/components/SortByTopButton.tsx [hunk 5]
```diff
@@ -1,35 +1,22 @@
...
-
 const SortByTopButton: React.FC<Props> = ({ sortedPosts }) => {
   const { button } = useStyles();
   const dispatch = useDispatch();
-  const posts = useSelector((state: StatePosts) => state.posts.posts);
 
-  useEffect(() => {
-    fetch("https://starforum.herokuapp.com/api/v1/posts")
+  const sortByTop = () => {
+    fetch("https://starforum.herokuapp.com/api/v1/posts?sort=most-upvotes")
```
Comment (vkravchuk2) [inline src/screens/main/components/SortByTopButton.tsx:17]: И здесь

**Review by vkravchuk2 — APPROVED** [review]

**Review by attiaahmedtarek — APPROVED** [review]

**Pull request merged**